pkg/controller/master-controller-manager/kubeone/controller.go

```diff
@@ -231,91 +229,220 @@ func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (
 	}
 
-	wantedVersion := cluster.Versions.Kubernetes
-	// check if a upgrade pod already exists
+	return cluster.Versions.Kubernetes, nil
+}
+
+func (r *reconciler) checkPodStatusIfExists(ctx context.Context,
+	log *zap.SugaredLogger,
+	externalCluster *kubermaticv1.ExternalCluster,
+	action string) error {
 	pod := &corev1.Pod{}
-	err = r.Get(ctx, ctrlruntimeclient.ObjectKey{
-		Name:      KubeOneUpgradePod,
+	var podName string
+	if action == UpgradeControlPlaneAction {
+		podName = KubeOneUpgradePod
+	} else if action == MigrateControlPlaneAction {
+		podName = KubeOneMigratePod
+	}
+
+	err := r.Get(ctx, ctrlruntimeclient.ObjectKey{
+		Name:      podName,
 		Namespace: kubernetesprovider.GetKubeOneNamespaceName(externalCluster.Name),
 	},
 		pod,
 	)
 	if err != nil {
-		if kerrors.IsNotFound(err) {
-			if externalCluster.Spec.CloudSpec.KubeOne.ClusterStatus.Status != kubermaticv1.StatusError {
-				if currentVersion != wantedVersion {
-					log.Debugw("Upgrading kubeone cluster", "from", currentVersion, "to", wantedVersion)
-					if err := r.upgradeCluster(ctx, log, externalCluster); err != nil {
-						return reconcile.Result{}, err
-					}
-				}
-			} else {
-				return reconcile.Result{}, nil
-			}
-		} else {
-			return reconcile.Result{}, err
+		if !kerrors.IsNotFound(err) {
+			return err
 		}
 	} else {
-		if pod.Status.Phase == corev1.PodPending || pod.Status.Phase == corev1.PodRunning {
-			return reconcile.Result{RequeueAfter: time.Second * 10}, nil
-		} else {
-			err = r.checkPodStatus(ctx, log, pod, externalCluster)
-			if err != nil {
-				return reconcile.Result{}, err
-			}
+		err = r.checkPodStatus(ctx, log, pod, externalCluster, action)
+		if err != nil {
+			return err
 		}
 	}
+	return nil
+}
+
+func (r *reconciler) reconcile(ctx context.Context, log *zap.SugaredLogger, externalCluster *kubermaticv1.ExternalCluster) (reconcile.Result, error) {
+	if err := r.initiateImportAction(ctx, log, externalCluster); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	clusterStatus := externalCluster.Spec.CloudSpec.KubeOne.ClusterStatus.Status
+	externalClusterProvider, err := kubernetesprovider.NewExternalClusterProvider(r.ImpersonationClient, r.Client)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	manifestSecret := &corev1.Secret{}
+	if err := r.Get(ctx, types.NamespacedName{Namespace: externalCluster.Spec.CloudSpec.KubeOne.ManifestReference.Namespace, Name: externalCluster.Spec.CloudSpec.KubeOne.ManifestReference.Name}, manifestSecret); err != nil {
+		return reconcile.Result{}, fmt.Errorf("can not retrieve kubeone manifest secret: %w", err)
+	}
+	currentManifest := manifestSecret.Data[resources.KubeOneManifest]
+
+	_, err = r.upgradeAction(ctx, log, externalCluster, externalClusterProvider, currentManifest, clusterStatus)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if err = r.checkPodStatusIfExists(ctx, log, externalCluster, UpgradeControlPlaneAction); err != nil {
```

Comment:
How about giving this function the pod name instead of the action name? Then it would not have to have some if-else-elseif-else-if-if-else chain :)